### PR TITLE
Eval symlinks on XDG_RUNTIME_DIR

### PR DESCRIPTION
**Pull request by rhatdan**
Partial Fix for https://github.com/containers/podman/issues/14606

[NO NEW TESTS NEEDED]

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

File: pkg/rootless/rootless_linux.c
Comment (mheon) [inline pkg/rootless/rootless_linux.c:362]: Whitespace here looks a bit wierd

**Comment by mheon** [comment]
Code LGTM, I'm good to merge whenever the tests start passing

**Comment by rhatdan** [comment]
@edsantiago Any idea what is breaking the Windows man page generator?  I don't see how this PR is causing this error?

**Comment by edsantiago** [comment]
The error I see is:
```
docs: no doc file nor link docs/build/man/podman-.1 for 'podman '
```
The space looks suspicious but I don't know this code and have no idea what it means. Am looking into it now.

The other weird thing I see is:
```
lstat podman_tmp_5mgn/podman-run--1/bus: no such file or directory
```
I think that's related, because the string `lstat` does not appear on a successful CI run of this step. Continuing to investigate, am posting this in case it triggers a memory in anyone.

**Comment by edsantiago** [comment]
Okay..... there's a weird error involving "bus", and your PR adds the string "bus". I think there's a connection. Maybe `XDG_WHATEVER` is unset on Macs?

**Comment by edsantiago** [comment]
I see your new push but I don't think that's going to fix it: `XDG_RUNTIME_DIR` gets set complicatedly in one of the config.go files. I'm looking right now at https://github.com/containers/podman/blob/f6053cea8d220b8b7af9a9c92b65d9b9d7877eaf/pkg/util/utils_supported.go#L42-L43

...which has the string "podman-run-%s" which closely matches the "podman-run--1" in the lstat error (except, wtf, is "-1" a valid UID in Mac-land???). The other part, `podman_tmp_5mgn`, is set in Makefile: that's `$TMPDIR`. So XDG is getting set to a default value in TMPDIR, and something is then failing later. But I can't reproduce it on my end no matter what I use for TMPDIR and XDG. Still looking.

**Comment by edsantiago** [comment]
This patch (against main, not against your PR) fixes it. Obviously there may be a better solution:
```diff
diff --git a/cmd/podman/registry/config.go b/cmd/podman/registry/config.go
index a118fdc4d..0f6028888 100644
--- a/cmd/podman/registry/config.go
+++ b/cmd/podman/registry/config.go
@@ -105,7 +105,12 @@ func setXdgDirs() error {
        if _, found := os.LookupEnv("DBUS_SESSION_BUS_ADDRESS"); !found {
                sessionAddr := filepath.Join(os.Getenv("XDG_RUNTIME_DIR"), "bus")
                if _, err := os.Stat(sessionAddr); err == nil {
-                       os.Setenv("DBUS_SESSION_BUS_ADDRESS", "unix:path="+sessionAddr)
+                       dereferenced, err := filepath.EvalSymlinks(sessionAddr)
+                       if err != nil {
+                               return err
+                       }
+
+                       os.Setenv("DBUS_SESSION_BUS_ADDRESS", "unix:path="+dereferenced)
                }
        }
 ```

File: cmd/podman/registry/config.go
Comment (edsantiago) [inline cmd/podman/registry/config.go:107]: This isn't quite the same as mine. Even if you get it to compile, the logic is different: in my approach, I first assemble `XDG/bus`, _then_ check if that exists, _then_ deref links.

`XDG` will always be set. That's some early `config` hackery that @giuseppe added years ago. What does not exist on macs is `XDG + bus`.

File: libpod/reset.go
Comment (edsantiago) [inline libpod/reset.go:161]: Again, I think this is misleading because `XDG` is always, always going to be set very early on by `config` code.

File: pkg/systemd/dbus.go
Comment (edsantiago) [inline pkg/systemd/dbus.go:133]: And, like above, I think you need an existence check before you `EvalSymlinks`.

File: libpod/reset.go
Comment (rhatdan) [reply libpod/reset.go:161]: There is no change here if XDG_RUNTIME_DIR is not set or is set to "", Current code would not fail if XDG_RUNTIME_DIR=="" new code would, so I add his check.

File: pkg/systemd/dbus.go
Comment (rhatdan) [reply pkg/systemd/dbus.go:133]: What would the check do if it does not exists?

Still call
                        return godbus.Dial(fmt.Sprintf("unix:path=%s", path))
?

 If the file does not exists EvalSymlinks will return that it does not exists.

Comment (edsantiago) [reply pkg/systemd/dbus.go:133]: I don't know enough about this `systemd/private` path to comment. But:
* old code: just returns a string with the path, whether the path exists or not (leave it to caller to deal with problems)
* new code: barfs if path does not exist. This is new behavior. I don't know if it's good new behavior or bad new behavior... but it's a change.

Comment (rhatdan) [reply pkg/systemd/dbus.go:133]: I think the old code will barf when godbus.Dial attempts to connect to a non existing socket
The new code will report the non existing socket earlier, as I read it.

File: pkg/util/utils_supported.go
Comment (edsantiago) [inline pkg/util/utils_supported.go:33]: Kind of made me nervous to return early, but AFAICT it's perfectly fine, no code paths are affected, and this actually seems cleaner. PR LGTM but this has too many ramifications for me.

**Comment by rhatdan** [comment]
@containers/podman-maintainers PTAL

**Review by Luap99 — CHANGES_REQUESTED** [review]
Please add the test from my PR https://github.com/containers/podman/pull/14668 to make sure it actually fixes the issue.

**Comment by edsantiago** [comment]
Thanks for the test, @Luap99. Here's a version that doesn't depend on `/var/run` existing and being a symlink.
```bash
@test "podman rootless netns works when XDG_RUNTIME_DIR includes symlinks" {
    # regression test for https://github.com/containers/podman/issues/14606
    is_rootless || skip "only meaningful for rootless"

    # Create a tmpdir symlink pointing to /run, and use it briefly
    ln -s /run $PODMAN_TMPDIR/run
    local tmp_run=$PODMAN_TMPDIR/run/user/$(id -u)
    test -d $tmp_run || skip "/run/user/MYUID unavailable"

    # This 'run' would previously fail with:
    #    IPAM error: failed to open database ....
    XDG_RUNTIME_DIR=$tmp_run run_podman run --network bridge --rm $IMAGE ip a
    assert "$output" =~ "eth0"
}
```
(warning: if you run this test with podman @ main, or podman without this patch, you destroy everything and need to run `podman system reset` before you can run podman again. No idea why, don't really care).

**Comment by rhatdan** [comment]
@Luap99 @edsantiago PTAL

**Pull request merged**
